### PR TITLE
`api_gateway_rest_api`: Adds parameter fail_on_warnings [rebased]

### LINKED ISSUE #19154: Support including failonwarnings when API Gateways are imported from OpenAPI templates in `aws_api_gateway_rest_api`
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Note: This is a duplicate of #17971 which seems to have been closed with only support for `aws_apigatewayv2_api`. The purpose of this issue is to get it fixed for `aws_api_gateway_rest_api` as well.

There is already a PR in #18397 to fix this issue.

When creating an API Gateway from an OpenAPI template, the default behavior in AWS is to silently ignore warnings on import. This can lead to a situation where a Terraform deployment of the API Gateway is successful without any changes actually being deployed (e.g. a typo in a schema component causing routes to not be added).

To fix this, the [aws_api_gateway_rest_api](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/api_gateway_rest_api) resources should support adding failonwarnings to the API call that creates the resource.

### New or Affected Resource(s)

* [aws_api_gateway_rest_api](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/api_gateway_rest_api)

### Potential Terraform Configuration

```hcl
resource aws_api_gateway_rest_api example {
  name = "example-http-api"
  body = "..."
  fail_on_warnings = true
}
```

### References

* [failonwarnings](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-import-api-errors-warnings.html)
* [API reference](https://docs.aws.amazon.com/apigateway/api-reference/link-relation/restapi-import/)
* [Cloudformation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigateway-restapi.html#cfn-apigateway-restapi-failonwarnings)

### LINKED ISSUE #22674: aws_api_gateway_rest_api should provide warning output (#4274 repeat)
This is a repeat of #4274 as a bug, because that's what I think it is.  There are many semantic errors which can be made when importing or updating an OpenAPI/Swagger spec (using the `body` argument), and these are never surfaced by the `aws_api_gateway_rest_api` resource.

This is because the provider calls [APIGateway.PutRestApi](https://docs.aws.amazon.com/sdk-for-go/api/service/apigateway/#APIGateway.PutRestApi) without specifying FailOnWarnings (which defaults to false) in the [PutRestApiInput](https://docs.aws.amazon.com/sdk-for-go/api/service/apigateway/#PutRestApiInput):

- https://github.com/hashicorp/terraform-provider-aws/blob/main/internal/service/apigateway/rest_api.go#L217
- https://github.com/hashicorp/terraform-provider-aws/blob/main/internal/service/apigateway/rest_api.go#L598

This violates the principle of least astonishment because the body reported in the resource output is what was provided to the resource, but it is rejected by the AWS APIs and never applied to the API Gateway.  This means it's impossible to debug using only terraform.  (I ended up expanding the template manually into a file and attempting to import this into API Gateway to debug it.)

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

```
Terraform v1.0.10
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v3.72.0
+ provider registry.terraform.io/hashicorp/local v2.1.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
+ provider registry.terraform.io/hashicorp/template v2.2.0
```

### Affected Resource(s)

* aws_api_gateway_rest_api

### Terraform Configuration Files

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

- [main.tf](https://gist.github.com/paulgear/7bce944430fe48011a3e3aa598915bfe)
- [api.json](https://gist.github.com/paulgear/59bf1dc41b7924969a5e6a30a9709c8f)
- [default_policy.json](https://gist.github.com/paulgear/df549ef38cedc9d6238bd74c1eaaa3b9)
- [Documentation from Auth0 about the solution](https://auth0.com/docs/customize/integrations/aws/aws-api-gateway-custom-authorizers), for additional context, if needed

### Debug Output

n/a

### Panic Output

n/a

### Expected Behavior

- We should be able to see why our API body import is failing.

### Actual Behavior

- terraform plan/apply completes successfully with no errors, but our API body changes were not imported.

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

- The current [AWS API docs](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-swagger-extensions-authorizer.html) are out-of-date with respect to OpenAPI 3.x (specifically with respect to the [naming and placement of `securityDefinitions`](https://swagger.io/docs/specification/authentication/), and working from the examples provided there results in a non-functional API gateway authorizer, which makes it all the more important that terraform provide a way to debug it.

### References

* #4274
* I'll attempt a draft PR on this shortly, but as a non-native golang speaker, I expect it will need some assistance to make it through the PR process.

### PR DESCRIPTION
**Pull request by fullammo**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #19154.
Closes #22674.
Closes https://github.com/hashicorp/terraform-provider-aws/pull/18397.

This PR is an up to date version of [PR#18397](https://github.com/hashicorp/terraform-provider-aws/pull/18397), which complies to the [mentioned](https://github.com/hashicorp/terraform-provider-aws/pull/18397#issuecomment-946903217) refactoring changes that had to be addressed.

Output from acceptance testing:
```shell
% make testacc TESTS=TestAccAPIGatewayRestApi_Fail_On_Warnings PKG=apigateway
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/apigateway/... -v -count 1 -parallel 20 -run='TestAccAPIGatewayRestApi_Fail_On_Warnings' -timeout 180m
=== RUN   TestAccAPIGatewayRestApi_Fail_On_Warnings
=== PAUSE TestAccAPIGatewayRestApi_Fail_On_Warnings
=== CONT  TestAccAPIGatewayRestApi_Fail_On_Warnings
--- PASS: TestAccAPIGatewayRestApi_Fail_On_Warnings (58.65s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/apigateway 61.240s
```
@johanwiren This was the easiest way to resolve your merge conflicts, I hope it's not a problem. 😄

### CODE REVIEW

**Review by AdamTylerLynch — APPROVED** [review]
LGTM 🚀

**Comment by fullammo** [comment]
Awesome! Could this change be included in one of the upcoming releases? Thanks in advance!

**Comment by fullammo** [comment]
There were merge conflicts, i resolved them, just to make the merge a bit smoother  :1st_place_medal:

**Comment by michael-rivera-florencehc-com** [comment]
Any updates on this? We can use this as well, and it would help a bunch?

**Comment by thxmike** [comment]
This would be a good feature to implement to prevent any deployment issues. Please merge this in

**Comment by DaniBencz** [comment]
It would be an improvement to the current behavior that 'swallows' warnings. Our team could benefit from it.

**Comment by njovanovic-fhc** [comment]
Hey team, any update on this? Our team can utilise this as well...

**Comment by somaSzolyak** [comment]
Hello!

This feature would help our team as well!

File: internal/service/apigateway/rest_api.go [hunk 1]
```diff
@@ -55,7 +55,10 @@ func ResourceRestAPI() *schema.Resource {
 			},
 			"created_date": {
 				Type:     schema.TypeString,
-				Computed: true,
+
+			"fail_on_warnings": {
```
Comment (AdamTylerLynch) [inline internal/service/apigateway/rest_api.go:59]: It's important to note that the [failOnWarnings](https://docs.aws.amazon.com/apigateway/latest/api/API_PutRestApi.html) parameter only applies to AWS's PutRestApi() method when using an OpenAPI definition. This is an update of an existing API only.

Terraform's implementation of this resource is slightly divergent from the intended use of the API where Terraform first creates a REST API (without an OpenAPI definition) and then if a OpenAPI definition is supplied immediately updates the API using the OpenAPI definition.

File: website/docs/r/api_gateway_rest_api.html.markdown [hunk 2]
```diff
@@ -229,6 +229,7 @@ The following arguments are supported:
 * `endpoint_configuration` - (Optional) Configuration block defining API endpoint configuration including endpoint type. Defined below.
 * `minimum_compression_size` - (Optional) Minimum response size to compress for the REST API. Integer between `-1` and `10485760` (10MB). Setting a value greater than `-1` will enable compression, `-1` disables compression (default). If importing an OpenAPI specification via the `body` argument, this corresponds to the [`x-amazon-apigateway-minimum-compression-size` extension](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-openapi-minimum-compression-size.html). If the argument value (_except_ `-1`) is provided and is different than the OpenAPI value, the argument value will override the OpenAPI value.
 * `name` - (Required) Name of the REST API. If importing an OpenAPI specification via the `body` argument, this corresponds to the `info.title` field. If the argument value is different than the OpenAPI value, the argument value will override the OpenAPI value.
+* `fail_on_warnings` - (Optional) Whether warnings while API Gateway is creating or updating the resource should return an error or not. Defaults to `false`
```
Comment (AdamTylerLynch) [inline website/docs/r/api_gateway_rest_api.html.markdown:232]: In our documentation, arguments are listed in alphabetical order. This argument should come after `endpoint_configuration` and before `minimum_compression_size`.

Comment (AdamTylerLynch) [inline website/docs/r/api_gateway_rest_api.html.markdown:232]: This argument is only applicable when a body (OpenAPI definition) argument is supplied.

```suggestion
* `fail_on_warnings` - (Optional) Failure mode of the PutRestApi operation when importing an OpenAPI specification via the body argument (create or update operation). Indicates whether to rollback the API update (true) or not (false) when a warning is encountered. Defaults to `false`.
```

**Review by AdamTylerLynch — COMMENTED** [review]
AWS's API requires first creating a REST API and then updating that API if an OpenAPI spec is supplied.

Terraform's implementation of this resource is aligned to that intended use, where Terraform first creates a REST API (without an OpenAPI definition) and then if a OpenAPI definition is supplied immediately updates the API using the OpenAPI definition.

A failOnWarning during `create` would most likely result in an empty REST API, as the rollback would rollback the OpenAPI application step.

A failOnWarning during `update` would result in the expected behavior, but would leave the Terraform state in an inconsistent state. My hypothesis is the author saw this inconsistent state when testing, and this is why they removed the computed field from `create_date`.

Implementation of this `failOnWarning` feature introduces to many permutations of the potential resource and state due to the AWS APIs requirements, and this would require significant more AccTests to provide surety on permutation coverage.

**Pull request closed without merging**